src/Interpreters/Cache/FileSegment.h

```diff
@@ -305,46 +309,33 @@ friend class StorageSystemFilesystemCache;
 
     RemoteFileReaderPtr remote_file_reader;
     LocalCacheWriterPtr cache_writer;
-    bool detached_writer = false;
 
     /// downloaded_size should always be less or equal to reserved_size
-    size_t downloaded_size = 0;
-    size_t reserved_size = 0;
-
-    /// global locking order rule:
-    /// 1. cache lock
-    /// 2. segment lock
+    std::atomic<size_t> downloaded_size = 0;
+    std::atomic<size_t> reserved_size = 0;
+    mutable std::mutex download_mutex;
 
-    mutable std::mutex mutex;
+    mutable FileSegmentGuard segment_guard;
+    KeyMetadata * key_metadata;
     std::condition_variable cv;
 
-    /// Protects downloaded_size access with actual write into fs.
-    /// downloaded_size is not protected by download_mutex in methods which
-    /// can never be run in parallel to FileSegment::write() method
-    /// as downloaded_size is updated only in FileSegment::write() method.
-    /// Such methods are identified by isDownloader() check at their start,
-    /// e.g. they are executed strictly by the same thread, sequentially.
-    mutable std::mutex download_mutex;
-
     Key file_key;
+    const std::string file_path;
     FileCache * cache;
 
     Poco::Logger * log;
 
-    /// "detached" file segment means that it is not owned by cache ("detached" from cache).
-    /// In general case, all file segments are owned by cache.
-    bool is_detached = false;
-    bool is_completed = false;
-
-    bool is_downloaded = false;
+    /// Does the file segment have completed state?
```

Comment:
Please add file segment states which are considered as completed

Reply:
Actually the flag can be replaced with check that `download_state` is equal one of final states. So, consider to remove it

Reply:
This is not really that straightforward, because PARTIALLY_DOWNLOADED_NO_CONTINUATION is neither a completed state nor a non-completed state, e.g. this state can be a final state but it can also be non-final state. It becomes final after a copy of this file segment was resized to the downloaded and put to cache and current made detached, but valid - serving as a detached read-only view to the actual resized file segment.

But I agree that it is nice to get rid of this variable, so I made some hack, which I think is acceptable. You will see it in the next commit (haven't pushed yet).

Reply:
Made it better: [b8e958d](https://github.com/ClickHouse/ClickHouse/pull/44985/commits/b8e958df07ac3a88b229e25fbcd08467f8701e95).